factcast-snapshotcache-jdbc/src/main/java/org/factcast/core/snap/jdbc/JdbcSnapshotCache.java

```diff
@@ -0,0 +1,182 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.factcast.core.snap.jdbc;
+
+import com.google.common.collect.Sets;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.time.LocalDate;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+import javax.sql.DataSource;
+import lombok.NonNull;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.factcast.core.snap.Snapshot;
+import org.factcast.core.snap.SnapshotId;
+import org.factcast.factus.snapshot.SnapshotCache;
+
+@Slf4j
+public class JdbcSnapshotCache implements SnapshotCache {
+  public final String queryStatement;
+  public final String mergeStatement;
+  public final String deleteStatement;
+  public final String cleanupStatement;
+  private final DataSource dataSource;
+
+  public JdbcSnapshotCache(JdbcSnapshotProperties properties, DataSource dataSource) {
+    this.dataSource = dataSource;
+
+    queryStatement =
+        "SELECT * FROM " + properties.getSnapshotsTableName() + " WHERE key = ? AND uuid = ?";
+
+    mergeStatement =
+        "MERGE INTO "
+            + properties.getSnapshotsTableName()
+            + " USING (VALUES (?, ?, ?, ?, ?, ?)) as new (_key, _uuid, _last_fact_id, _bytes, _compressed, _last_accessed)"
+            + " ON key=_key AND uuid=_uuid"
+            + " WHEN MATCHED THEN"
+            + " UPDATE SET last_fact_id=_last_fact_id, bytes=_bytes, compressed=_compressed, last_accessed=_last_accessed"
+            + " WHEN NOT MATCHED THEN"
+            + " INSERT VALUES (_key, _uuid, _last_fact_id, _bytes, _compressed, _last_accessed)";
+
+    deleteStatement =
+        "DELETE FROM " + properties.getSnapshotsTableName() + " WHERE key = ? AND uuid = ?";
+    cleanupStatement =
+        "DELETE FROM " + properties.getSnapshotsTableName() + " WHERE last_accessed < ?";
+
+    boolean snapTableExists = doesTableExist(properties.getSnapshotsTableName());
+
+    if (!snapTableExists) {
+      throw new IllegalStateException(
+          "Snapshots table does not exist: " + properties.getSnapshotsTableName());
+    } else {
+      validateColumns(properties.getSnapshotsTableName());
+    }
+
+    if (properties.getDeleteSnapshotStaleForDays() > 0) {
+      Timer timer = new Timer();
+      timer.scheduleAtFixedRate(
+          deleteStaleSnapshotsTimerTask(dataSource, properties.getDeleteSnapshotStaleForDays()),
+          0,
+          TimeUnit.DAYS.toMillis(1));
+    } else {
+      log.info("Snapshot cleanup is disabled");
+    }
+  }
+
+  private TimerTask deleteStaleSnapshotsTimerTask(DataSource dataSource, int staleForDays) {
```

Comment:
minor: wouldn't it be more intuitive to have a local subclass of TimeTask instead?